toqito/matrices/cyclic_permutation.py

```diff
@@ -0,0 +1,61 @@
+"""Generate a cyclic permutation matrix."""
+
+import numpy as np
+
+
+def cyclic_permutation_matrix(n: int) -> np.ndarray:
+    r"""Create the cyclic permutation matrix for a given dimension :code:`n` :cite:`WikiCyclicPermutation`.
+
+    This function creates a cyclic permutation matrix which is a special type of square matrix
+    that represents a cyclic permutation of its rows.
+
+    The permutation can be written in cycle notation and two-line notation as:
+
+    .. math::
+        \begin{align*}
+        \begin{aligned}
+        &\begin{matrix}
+        (1 & 4 & 6 & 8 & 3 & 7)(2)(5)
+        \end{matrix}\\
+        &= \begin{pmatrix}
+        1 & 2 & 3 & 4 & 5 & 6 & 7 & 8 \\
+        4 & 2 & 7 & 6 & 5 & 8 & 1 & 3
+        \end{pmatrix}\\
+        &= \begin{pmatrix}
+        1 & 4 & 6 & 8 & 3 & 7 & 2 & 5 \\
+        4 & 6 & 8 & 3 & 7 & 1 & 2 & 5
+        \end{pmatrix}
+        \end{aligned}
+        \end{align*}
+
+    Examples
+    ==========
+    Generate various cyclic permutation matrices.
+
+    >>> from this_module import cyclic_permutation_matrix
+    >>> n = 4
+    >>> cyclic_permutation_matrix = cyclic_permutation_matrix(n)
+    >>> print(cyclic_permutation_matrix)
+    ([[0, 0, 0, 1],
+    [1, 0, 0, 0],
+    [0, 1, 0, 0],
+    [0, 0, 1, 0]])
+    ...
+
+    References
+    ==========
+    .. bibliography::
+        :filter: docname in docnames
+
+    :param n: int
+        The number of rows and columns in the cyclic permutation matrix.
+
+    :return:
+        A NumPy array representing a cyclic permutation matrix of dimension :code:`n x n`.
+        Each row of the matrix is shifted one position to the right in a cyclic manner,
+        creating a circular permutation pattern.
+    """
+    p_mat = np.zeros((n, n), dtype=int)
+    np.fill_diagonal(p_mat[1:], 1)
+    p_mat[0, -1] = 1
+    return p_mat
```

Comment:
Does the function allow a user to choose the permutation order? I see here that the rows of a  `4 x 4` identity matrix are moved around as `1 -> 2, 2 -> 3, 3-> 4, 4 -> 1`

But you could also choose a different permutation order here.

Reply:
I was actually think about this topic yesterday and reviewing what Wiki had on it.

As of right now, we only have one permutation order.Tomorrow I'll review this piece again and then explore the idea of giving the user the ability to select a different order.

Reply:
If we want the ability for the user to choose between Successive application vs. Fixed points we could do something like:

```python
def cyclic_permutation_matrix(n: int, k: int = 1) -> np.ndarray:
    p_mat = np.zeros((n, n), dtype=int)
    np.fill_diagonal(p_mat[1:], 1)
    p_mat[0, -1] = 1

    result_mat = np.linalg.matrix_power(p_mat, k) # lets raise the matrix to the power of k
    return result_mat
```

With this, a user could decide which order they want to use:

Fixed points
```python
n = 4
cyclic_permutation_matrix = cyclic_permutation_matrix(n)
print(cyclic_permutation_matrix)
```

Result
```
[[0 0 0 1]
 [1 0 0 0]
 [0 1 0 0]
 [0 0 1 0]]
```

Successive application 
```python
n = 4
k = 3
cyclic_permutation_matrix = cyclic_permutation_matrix(n, k)
print(cyclic_permutation_matrix)
``` 

Result 
```
[[0 1 0 0]
 [0 0 1 0]
 [0 0 0 1]
 [1 0 0 0]]
```

@vprusso I'm curious if you have any preference on this. I think giving a user the ability to choose a different order would be a nice touch.

Reply:
Hi @Bchass, 

First of all, my apologies for my delayed response to your comment (the holidays were a bit busier than expected for me). 

W.r.t. your question, I think that's a nice way to make this function both more general and also more explicit. If the proper documentation can be added to describe this in the docstring, I am in full support of your idea. Nice suggestion!